doc/accerciser.1
.TH ACCERCISER 1 "27 August 2007"
.SH NAME
ACCERCISER \- an interactive accessibility explorer.
.SH SYNOPSIS
.B accerciser
[\-\-version] [\-\-help]
.SH DESCRIPTION
This manual page documents briefly
.B accerciser\fP.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBaccerciser\fP is an interactive Python accessibility explorer for the GNOME
desktop. It uses AT-SPI to inspect and control widgets, allowing you to
check if an application is providing correct information to assistive
technologies and automated test frameworks. Accerciser has a simple
plug-in framework which you can use to create custom views of
accessibility information.
.SH OPTIONS
.B \-\-version
Print the version number.
.TP
.B \-?, \-\-help
Show summary of options.
.SH CONFIGURATION
\fBmetacity\fP configuration can be found under \fIPreferences\fP->\fIWindows\fP and \fIPreferences\fP->\fIKeyboard Shortcuts\fP on the menu-panel. Advanced configuration can be achieved directly through gconf editing (gconf-editor or gconftool-2).
.SH SEE ALSO
Comprehensive \fBaccerciser\fP documentation could be found under the Help menu in \fBaccerciser\fP.
.SH AUTHOR
The original manual page was written by Eitan Isaacson <eitan@ascender.com>.


NEWS
==============
Version 0.1.91
==============

What's New
==========
- Added a manpage.

New And Updated Translations
============================
- Hendrik Richter (de)
- Adam Weinberger (en_CA)
- Ilkka Tuohela (fi)
- Gabor Kelemen (hu)
- Takeshi AIHANA (ja)
- Jovan Naumovski (mk)
- Kjartan Maraas (nb)
- Luiz Armesto (pt_BR)
- Duarte Loreto (pt)
- Daniel Nylander (sv)

==============
Version 0.1.90
==============

Bugs Fixed
==========
- Fixed some i18n issues.
- Fixed GTK+ 2.10 backward incompatibility, it has since been fixed
upstream (bug #460071).

New And Updated Translations
============================
- Joan Duran (ca)
- Jorge González (es)
- Ankit Patel (gu)
- Daniel Nylander (sv)
- Kiran Chandra (te)

=============
Version 0.1.6
=============

Bugs Fixed
==========
- Accerciser now blocks SIGSTP so the entire desktop won't hang (bug
#457965). Thanks Simos Xenitellis!
- Fixed bookmark management.
- Other tiny bug fixes.

New And Updated Translations
============================
- Runa Bhattacharjee (bn_IN)
- Simos Xenitellis (el)
- Stéphane Raimbault (fr)
- Rajesh Ranjan (hi)
- Takeshi AIHANA (ja)
- Raivis Dejus (lv)
- Matej Urbančič (sl)
- Daniel Nylander (sv)

=============
Version 0.1.5
=============

New Features
============
- Single plugin view mode, good for screen readers.
- Moved core hotkeys to a plugin called "Quick select".
- All plugins now scroll down with the focus.
- Updated English documentation.

New And Updated Translations
============================
- Jorge González (es)
- Stéphane Raimbault (fr)
- Ankit Patel (gu)
- Yannig MARCHEGAY (oc)
- Clytie Siddall (vi)

=============
Version 0.1.4
=============

New Features
============
- Accessible treeview's loading of child nodes is now decoupled,
loading obscenely large amounts of child nodes does not freeze the
entire desktop.
- Removed main glade file, main UI definition is now built at run time.
- Implemented use of UIManager and Actions.
- Added bookmark capabilities.

New And Updated Translations
============================
- Alexander Shopov (bg)
- Joan Duran (ca)
- Dawa pemo (dz)
- Jorge González (es)
- Jovan Naumovski (mk)

=============
Version 0.1.3
=============

New Features
============
- Plugin management has been refactored, much easier to maintain and 
  understand.
- Source entirely documented with Epydoc.
- Accessible description and child count in interface viewer plugin.
- Image description and locale in interface viewer plugin.

New And Updated Translations
============================
- Joan Duran (ca)
- David Lodge (en_GB)
- Jorge González (es)
- Robert-André Mauchin (fr)
- Eitan Isaacson (he)
- Subhransu Behera (or)
- Daniel Nylander (sv)

=============
Version 0.1.2
=============

New Features
============
- Accerciser now uses pyatspi, the new and official python at-spi
library.
- The Interfacee viewer plugin has bee rewritten.
- Faster startup.
- New role icons.
- Plugin tab context menus.

New And Updated Translations
============================
- Djihed Afifi (ar)
- David Lodge (en_GB)
- Jorge González (es)
- Kjartan Maraas (nb)
- Yannig Marchegay (oc)
- A S Alam (pa)
- Daniel Nylander (sv)

=============
Version 0.1.1
=============

New Features
============
- Script recorder plugin that records UI interaction in to a Dogtail
compatable script.
- A script playback module for scripts recorded via the script
recorder plugin.
- Added ability to configure all global hotkeys in new preferences
dialog.
- Moved to a gconf backend for persisted settings.
- New plugin preferenced dialog is now accessible with new context menus.
- New role icons (James Stipp).

New And Updated Translations
============================
- Djihed Afifi (ar)
- yangka (dz)
- Ignacio Casal Quinteiro (gl)
- Eitan Isaacson (he)
- Daniel Nylander (sv)


src/lib/accerciser/accerciser.py
'''
Defines the main program classes for creating the GUI, drawing the blinking
selection rectangle, and maintaining information about the currently selected
accessible.

@author: Peter Parente
@author: Eitan Isaacson
@organization: IBM Corporation
@copyright: Copyright (c) 2006, 2007 IBM Corporation
@license: BSD

All rights reserved. This program and the accompanying materials are made 
available under the terms of the BSD which accompanies this distribution, and 
is available at U{http://www.opensource.org/licenses/bsd-license.php}
'''
import pygtk
pygtk.require('2.0')
import gtk
import gobject
import gtk.glade
import gtk.gdk
import os, sys, locale
from icons import getIcon
import os
import atk
import gnome
from bookmarks import BookmarkStore
from accessible_treeview import *
from node import Node
from plugin import PluginManager
from plugin import PluginView
from tools import Tools
from i18n import _, N_
import wnck
from gnome import program_get
import gconf
from hotkey_manager import HotkeyManager, HotkeyTreeView
import gconf
from about_dialog import AccerciserAboutDialog
from prefs_dialog import AccerciserPreferencesDialog
from main_window import AccerciserMainWindow
import ui_manager

GLADE_FILENAME = os.path.join(sys.prefix, 'share', 'accerciser', 'glade', 
                              'accerciser.glade')
if not os.path.exists(GLADE_FILENAME):
  GLADE_FILENAME = os.path.join(os.getcwd(), 'accerciser.glade')
  
GCONF_GENERAL = '/apps/accerciser/general'

class Main(Tools):
  '''
  Class for the main accerciser window. 
  '''
  COL_ACC = 4
  COL_FILLED = 2
  
  def __init__(self):
    '''
    Gets references to important widgets, establishes event handlers,
    configures the tree view, and initializes the tree with the contents of the
    desktop.
    '''
    # mark the root of this window with its PID so we can easily identify it
    # as this app
    root_atk = atk.get_root()
    root_atk.set_description(str(os.getpid()))

    self.node = Node()

    self.window = AccerciserMainWindow(self.node)
    self.window.connect('destroy', self._onQuit)

    # Start hotkey manager
    self.hotkey_manager = HotkeyManager()

    self.node.connect('accessible_changed', self._onAccesibleChange)

    self.bookmarks_store = BookmarkStore(self.node)

    # load plugins
    self.plugin_manager = \
        PluginManager(self.node, self.hotkey_manager,
                      self.window.pluginview1, self.window.pluginview2)

    # connect signal handlers and show the GUI in its initial state
    self.window.show_all()

    main_actions = gtk.ActionGroup('MainActions')
    ui_manager.uimanager.insert_action_group(main_actions, 0)
    main_actions.add_actions([
        ('Quit', gtk.STOCK_QUIT, None, 
         '<control>q', 'Quit Accerciser', self._onQuit),
        ('Preferences', gtk.STOCK_PREFERENCES, _('_Preferences...'),
         '<control>p', 'Show preferences', self._onShowPreferences),
        ('Contents', gtk.STOCK_HELP, _('_Contents'),
         'F1', 'View contents of manual', self._onHelp),
        ('About', gtk.STOCK_ABOUT, None,
         None, 'About Accerciser', self._onAbout)])

    for action_name, menu_path in [('Quit', ui_manager.FILE_MENU_PATH),
                                  ('Preferences', ui_manager.EDIT_MENU_PATH),
                                  ('Contents', ui_manager.HELP_MENU_PATH),
                                  ('About', ui_manager.HELP_MENU_PATH)]:
      action = main_actions.get_action(action_name)
      ui_manager.uimanager.add_ui(ui_manager.uimanager.new_merge_id(), 
                                  menu_path, action_name, action_name, 
                                  gtk.UI_MANAGER_MENUITEM, False)


    self.last_focused = None
    self.window.show_all()

  def run(self):
    '''
    Runs the app.
    '''
    # Tell user if desktop accessibility is disabled.
    self._showNoA11yDialog()
    try:
      pyatspi.Registry.start()
    except KeyboardInterrupt:
      self._shutDown()

  def _showNoA11yDialog(self):
    '''
    Shows a dialog with a relevant message when desktop accessibility seems to
    be disabled. If desktop accessibility is disabled in gconf, prompts the
    user to enable it.
    '''
    cl = gconf.client_get_default()
    if not cl.get_bool('/desktop/gnome/interface/accessibility'):
      message = _('Accerciser could not see the applications on your desktop.'
                  'You must enable desktop accessibility to fix this problem.'
                  'Do you want to enable it now?')
      dialog = gtk.MessageDialog(self.window,type=gtk.MESSAGE_ERROR,
                                 buttons=gtk.BUTTONS_YES_NO, 
                                 message_format=message)
      response_id = dialog.run()
      dialog.destroy()
      if response_id == gtk.RESPONSE_YES:
        cl = gconf.client_get_default()
        cl.set_bool('/desktop/gnome/interface/accessibility', True)
        dialog = gtk.MessageDialog(
          self.window,
          type=gtk.MESSAGE_INFO,
          buttons=gtk.BUTTONS_OK, 
          message_format=_('Note: Changes only take effect after logout.'))
        dialog.run()
        dialog.destroy()

  def _onAccesibleChange(self, node, acc):
    '''
    Callback for "accessible_changed" signal that is emitted by the L{Node}
    referenced as an instance variable.
    Updates the status bar with the path to the selected accessible.

    @param node: The node that emitted the signal.
    @type node: L{Node}
    @param acc: The new accessible that is referenced by the node.
    @type acc: L{Accessibility.Accessible}
    '''
    # Update status bar
    statusbar = self.window.statusbar
    context_id = statusbar.get_context_id('lineage')
    selection = self.window.treeview.get_selection()
    model, iter = selection.get_selected()
    if not iter:
      return
    path = map(str, model.get_path(iter))
    statusbar.pop(context_id)
    if len(path) > 1:
      statusbar.push(context_id, 'Path: '+' '.join(path[1:]))

  def _shutDown(self):
    '''
    Cleans up any object instances that need explicit shutdown.
    '''
    self.window.saveState()
    self.plugin_manager.close()

  def _onQuit(self, obj):
    '''
    Quits the app.

    @param obj: The object that emitted the signal that this callback caught.
    @type obj: L{gtk.Widget}
    '''
    self._shutDown()
    pyatspi.Registry.stop()
    
  def _onAbout(self, action):
    '''
    Shows the about dialog.

    @param widget: The widget that emitted the signal that this callback caught.
    @type widget: L{gtk.Widget}
    '''
    about = AccerciserAboutDialog()
    about.show_all()
    
  def _onHelp(self, action):
    '''
    Shows the help dialog.

    @param widget: The widget that emitted the signal that this callback caught.
    @type widget: L{gtk.Widget}
    '''
    gnome.help_display('accerciser.xml')
         
  def _onShowPreferences(self, action):
    '''
    Shows the preferences dialog.

    @param widget: The widget that emitted the signal that this callback caught.
    @type widget: L{gtk.Widget}
    '''
    plugins_view = self.plugin_manager.View()
    hotkeys_view = HotkeyTreeView(self.hotkey_manager)
    dialog = AccerciserPreferencesDialog(plugins_view, hotkeys_view)
    dialog.show_all()
    
